File: lib/src/main/java/com/sd/libcore/business/FBusiness.java
package com.sd.libcore.business;

import android.text.TextUtils;

import androidx.annotation.CallSuper;

import com.sd.lib.stream.FStream;
import com.sd.libcore.business.stream.BSProgress;
import com.sd.libcore.business.stream.BSTipsCallback;

public abstract class FBusiness
{
    private String mTag;

    public FBusiness(String tag)
    {
        mTag = tag;
    }

    /**
     * 初始化
     */
    @CallSuper
    public void init()
    {
        FBusinessManager.getInstance().addBusiness(this);
    }

    /**
     * 返回当前业务类的标识
     *
     * @return
     */
    public final String getTag()
    {
        return mTag;
    }

    /**
     * 设置业务类标识
     *
     * @param tag
     */
    public final void setTag(String tag)
    {
        final String oldTag = mTag;
        if (!TextUtils.equals(oldTag, tag))
        {
            mTag = tag;
            onTagChanged(oldTag, tag);
        }
    }

    /**
     * 返回Http请求标识
     *
     * @return
     */
    public String getHttpTag()
    {
        return toString();
    }

    /**
     * 返回一个流代理对象
     *
     * @param clazz
     * @param <T>
     * @return
     */
    protected final <T extends FStream> T getStream(Class<T> clazz)
    {
        return new FStream.ProxyBuilder().setTag(getTag()).build(clazz);
    }

    public final BSProgress getProgress()
    {
        return getStream(BSProgress.class);
    }

    public final BSTipsCallback getTipsCallback()
    {
        return getStream(BSTipsCallback.class);
    }

    /**
     * 标识变化回调
     *
     * @param oldTag
     * @param newTag
     */
    protected void onTagChanged(String oldTag, String newTag)
    {
    }

    /**
     * 销毁
     */
    @CallSuper
    public void onDestroy()
    {
        FBusinessManager.getInstance().removeBusiness(this);
    }
}

File: lib/src/main/java/com/sd/libcore/business/FBusiness.kt
package com.sd.libcore.business

import android.text.TextUtils
import androidx.annotation.CallSuper
import com.sd.lib.libkt.coroutine.scope.FCoroutineScope
import com.sd.lib.libkt.coroutine.scope.FMainScope
import com.sd.lib.stream.FStream
import com.sd.lib.stream.FStream.ProxyBuilder
import com.sd.lib.tag_view.FTagView
import com.sd.lib.tag_view.ITagView
import com.sd.libcore.business.stream.BSProgress
import com.sd.libcore.business.stream.BSTipsCallback

abstract class FBusiness : FTagView.Item {
    private val _mainScope by lazy { FMainScope() }

    @JvmOverloads
    constructor(businessTag: String? = null) {
        tag = businessTag
    }

    /**
     * 业务类标识
     */
    var tag: String?
        set(value) {
            val oldTag = field
            if (!TextUtils.equals(oldTag, value)) {
                field = value
                onTagChanged(oldTag, value)
            }
        }

    /**
     * 返回Http请求标识
     */
    open val httpTag: String
        get() = toString()

    /**
     * 协程作用域
     */
    val mainScope: FCoroutineScope
        get() = _mainScope

    /**
     * [BSProgress]
     */
    val progress: BSProgress
        get() = getStream(BSProgress::class.java)

    /**
     * [BSTipsCallback]
     */
    val tipsCallback: BSTipsCallback
        get() = getStream(BSTipsCallback::class.java)

    /**
     * 初始化
     */
    @CallSuper
    open fun init() {
        _mainScope.init()
    }

    /**
     * 创建[clazz]的流代理对象
     */
    protected fun <T : FStream> getStream(clazz: Class<T>): T {
        return ProxyBuilder().setTag(tag).build(clazz)
    }

    /**
     * 标识变化回调
     *
     * @param oldTag 旧标识
     * @param newTag 新标识
     */
    protected open fun onTagChanged(oldTag: String?, newTag: String?) {}

    /**
     * 销毁
     */
    @CallSuper
    open fun onDestroy() {
        _mainScope.destroy()
    }

    final override fun initItem(tagView: ITagView) {
        tag = tagView.viewTag
        init()
    }

    final override fun destroyItem() {
        onDestroy()
    }
}